### PR TITLE
Migrate to a new PWD API (part 2)

### PR DESCRIPTION
**Pull request by YizhePKU**
Refer to #12603 for part 1.

We need to be careful when migrating to the new API, because the new API has slightly different semantics (PWD can contain symlinks). This PR handles the "obviously safe" part of the migrations. Namely, it handles two specific use cases:

* Passing PWD into `canonicalize_with()`
* Passing PWD into `EngineState::merge_env()`

The first case is safe because symlinks are canonicalized away. The second case is safe because `EngineState::merge_env()` only uses PWD to call `std::env::set_current_dir()`, which shouldn't affact Nushell. The commit message contains detailed stats on the updated files.

Because these migrations touch a lot of files, I want to keep these PRs small to avoid merge conflicts.

### CODE REVIEW

**Comment by YizhePKU** [comment]
By the way, why does `EngineState::merge_env()` need to call `std::env::set_current_dir()`? From my understanding, Nushell maintains its own PWD, so the actual PWD of the Nushell process is not used, right?

**Comment by kubouch** [comment]
Yeah, it shouldn't be needed. It is safe to call it there because `merge_env()` is always called in a single thread, but we could also try removing it because it shouldn't be needed.  You can check git blame for that line to see if there is some story behind it, I suspect it is left over from older code.

**Comment by YizhePKU** [comment]
I traced it to https://github.com/nushell/engine-q/pull/587 (kinda funny that it's your PR). Quote:

> The std::env is updated every time we merge the delta into the engine state. This ensures parser and completions have up-to-date current directory. Writing to std::env during the delta merge should be safe since we don't do it in multiple threads.

I think nowadays the parser and the completions also get PWD from our APIs, so the `std::env::set_current_dir()` call should be safe to remove. I'll open a PR to remove it after this PR is merged.

File: benches/benchmarks.rs [hunk 1]
```diff
@@ -21,8 +21,8 @@ fn load_bench_commands() -> EngineState {
 }
 
 fn canonicalize_path(engine_state: &EngineState, path: &Path) -> PathBuf {
-    #[allow(deprecated)]
-    let cwd = engine_state.current_work_dir();
+    // We're running a benchmark. Using `unwrap()` is fine.
```
Comment (IanManske) [inline benches/benchmarks.rs:24]: Agreed, feel free to remove the comment.

**Review by IanManske — APPROVED** [review]
Thanks! Looks good.

**Comment by kubouch** [comment]
Haha, thanks for digging it up! Well, it's my fault. I think I moved it there because previously we were calling it in `cd` command etc. which was wreaking havoc, so I at least moved it to `merge_env()`. It might be that we were still figuring it out how to best handle the PWD issue. Seems like it can be safely removed, indeed.

The only instance where I can think of `std::env::set_current_dir()` being “useful” is if you wanted to manually inspect the Nushell process's memory to figure out its environment. In that case, after removing the call from `merge_env()`, the process environment would contain the directory the process was spawned in, and it would never change. That's a nice consistent behavior which we could document as well if we want to be really pedantic.

**Pull request merged**